### PR TITLE
Add option to skip first token during logits comparison

### PR DESCRIPTION
**Pull request by gagika**
# Description
Adds `--skip_first_token` flag to `tests/forward_pass_logit_checker.py`. This ignores the first token during logit comparison, preventing logit comparison failures due to high entropy at the first token which is sensitivity to numeric accuracy especially for MoE models with many experts.

# Tests
Verified locally on CPU. The script successfully ignored the initial token mismatch and passed the test criteria for the subsequent tokens using the new flag.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

**Comment by RissyRan** [comment]
Wondering if the 1st token doesn't match, will this affect following tokens?

**Comment by gagika** [comment]
> Wondering if the 1st token doesn't match, will this affect following tokens?

at first token entropy is quite high (a lot of plausible options), so large (and MoE) models can be sensitive to numerical accuracy issues (e.g. routing to different expert).

2nd, 3rd, and upcoming tokens are all conditioned on the previous tokens (including first), and have smaller entropy, (less sensitive to numerical issues).

File: tests/forward_pass_logit_checker.py [hunk 1]
```diff
@@ -297,12 +297,16 @@ def main(config, test_args):  # pylint: disable=W0621
...
-      golden_logits_slice = golden_logits[:token_size, :min_vocab_size]
+      train_logits_slice = full_train_logits[0, start_index:token_size, :min_vocab_size]
+      golden_logits_slice = golden_logits[start_index:token_size, :min_vocab_size]
+
       max_logging.log("\n[logits: token 2]")
-      max_logging.log(f"{golden_logits_slice[2]=}")
-      max_logging.log(f"{train_logits_slice[2]=}")
+      if train_logits_slice.shape[0] > 2:
+        max_logging.log(f"{golden_logits_slice[2]=}")
+        max_logging.log(f"{train_logits_slice[2]=}")
```
Comment (shuningjin) [inline tests/forward_pass_logit_checker.py:314]: nit: could you move this log inside if and update the index
```
if train_logits_slice.shape[0] > 2:
  max_logging.log(f"\n[logits: token {start_index+2}]")
```
same for `max_logging.log("\n[probability: token 1]")` in the below

File: tests/utils/forward_pass_logit_checker.py [hunk 2]
```diff
@@ -499,6 +509,13 @@ def main(config, test_args):  # pylint: disable=W0621
   parser.add_argument("--output_logits_path", type=str, required=False, default="")
   parser.add_argument("--gcs_output_logits_path", type=str, required=False, default="")
   parser.add_argument("--clip_logits_epsilon", type=float, required=False, default=None)
+  parser.add_argument(
+      "--skip_first_token",
```
Comment (shuningjin) [inline tests/utils/forward_pass_logit_checker.py:518]: "--skip_first_token" is only effective when we compare against pre-generated logit file, but not when run hf model on the fly (`run_hf_model=False`). 
- For clarity, might be good to raise not implemented error if `run_hf_model=false` and skip_first_token=true`
- same for "--clip_logits_epsilon"

**Review by shuningjin — APPROVED** [review]
Thanks for identifying the noise in the initial token prediction and adding the option to skip! LGTM.

**Comment by RissyRan** [comment]
> > Wondering if the 1st token doesn't match, will this affect following tokens?
> 
> at first token entropy is quite high (a lot of plausible options), so large (and MoE) models can be sensitive to numerical accuracy issues (e.g. routing to different expert).
> 
> 2nd, 3rd, and upcoming tokens are all conditioned on the previous tokens (including first), and have smaller entropy, (less sensitive to numerical issues).

Thanks for the explanation! Potentially you have some data points for Kimi large model in logits verification. My mental model is: the first token is the most volatile due to high entropy. However, because each token depends on the previous one, that initial "tweak" usually causes the 2nd and 3rd tokens to diverge further from the original path, not settle back down. Right?

**Comment by RissyRan** [comment]
> > > Wondering if the 1st token doesn't match, will this affect following tokens?
> > 
> > 
> > at first token entropy is quite high (a lot of plausible options), so large (and MoE) models can be sensitive to numerical accuracy issues (e.g. routing to different expert).
> > 2nd, 3rd, and upcoming tokens are all conditioned on the previous tokens (including first), and have smaller entropy, (less sensitive to numerical issues).
> 
> Thanks for the explanation! Potentially you have some data points for Kimi large model in logits verification. My mental model is: the first token is the most volatile due to high entropy. However, because each token depends on the previous one, that initial "tweak" usually causes the 2nd and 3rd tokens to diverge further from the original path, not settle back down. Right?

Discussed offline with @gagika! It makes more sense now. This could be an issue for large MoE models (like Kimi-K2), potentially related with numerics, even the tokens IDs and weights are fixed.

File: tests/utils/forward_pass_logit_checker.py (hunk 2 above)
Comment (gagika) [reply tests/utils/forward_pass_logit_checker.py:518]: `--skip_first_token` should work for both paths pre-generated and hf model.

I am raising error for --clip_logits_epsilon on hf run.

File: tests/forward_pass_logit_checker.py (hunk 1 above)
Comment (gagika) [reply tests/forward_pass_logit_checker.py:314]: done, good catch.
